src/sys/windows/selector.rs

```diff
@@ -1,596 +1,541 @@
+            let result = unsafe {
+                self.afd
+                    .poll(&mut self.poll_info, self.iosb.as_mut(), overlapped)
+            };
+            if let Err(e) = result {
+                if let Some(code) = e.raw_os_error() {
+                    if code == ERROR_IO_PENDING as i32 {
+                        /* Overlapped poll operation in progress; this is expected. */
+                    } else if code == ERROR_INVALID_HANDLE as i32 {
+                        /* Socket closed; it'll be dropped. */
+                        self.mark_delete();
+                        return Ok(());
+                    } else {
+                        return Err(e);
+                    }
+                }
+            }
+
+            self.poll_status = SockPollStatus::Pending;
+            self.pending_evts = self.user_evts;
         } else {
-            // The readiness queue is not empty, so do not block the thread.
-            timeout = Some(Duration::from_millis(0));
+            unreachable!();
         }
+        Ok(())
+    }
 
-        let ret = self.select2(events, timeout)?;
-
-        // Poll custom event queue
-        self.readiness_queue.poll(events);
-
-        // Return number of polled events
-        Ok(ret)
+    fn cancel(&mut self) -> io::Result<()> {
+        assert!(self.poll_status == SockPollStatus::Pending);
+        unsafe {
+            self.afd.cancel(self.iosb.as_mut())?;
+        }
+        self.poll_status = SockPollStatus::Cancelled;
+        self.pending_evts = 0;
+        Ok(())
     }
 
-    pub fn select2(&self, events: &mut Events, timeout: Option<Duration>) -> io::Result<()> {
-        trace!("select; timeout={:?}", timeout);
+    fn mark_delete(&mut self) {
+        if !self.delete_pending {
+            if self.poll_status == SockPollStatus::Pending {
+                drop(self.cancel());
+            }
 
-        // Clear out the previous list of I/O events and get some more!
-        events.clear();
+            self.delete_pending = true;
+        }
+    }
 
-        trace!("polling IOCP");
-        let n = match self.inner.port.get_many(&mut events.statuses, timeout) {
-            Ok(statuses) => statuses.len(),
-            Err(ref e) if e.raw_os_error() == Some(WAIT_TIMEOUT as i32) => 0,
-            Err(e) => return Err(e),
-        };
+    fn feed_event(&mut self) -> Option<Event> {
+        let mut afd_events = 0;
+        self.poll_status = SockPollStatus::Idle;
+        self.pending_evts = 0;
 
-        for status in events.statuses[..n].iter() {
-            // This should only ever happen from the waker.
-            if status.overlapped() as usize == 0 {
-                let token = Token(status.token());
-                if token == WAKE {
-                    continue;
-                }
-                events.events.push(Event::new(Ready::READABLE, token));
-                continue;
+        unsafe {
```

Comment:
Can you reduce this unsafe block and add a comment why it is safe?

Reply:
I cannot because the logic is connected. It is better to put this still.
```
// We use the status info in IO_STATUS_BLOCK to determine the socket poll status. It is unsafe to use a pointer of IO_STATUS_BLOCK.
```